### PR TITLE
[xml-model] XML Completion based on DTD/XML Schema by using xml-model

### LINKED ISSUE #698: [xml-model] XML Completion based on DTD/XML Schema by using xml-model
[xml-model] XML Completion based on DTD/XML Schema by using xml-model

### PR DESCRIPTION
**Pull request by angelozerr**
[xml-model] XML Completion based on DTD/XML Schema by using xml-model

Fixes #698

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
You can test this PR with this XML:

```
<?xml-model href="http://maven.apache.org/xsd/maven-4.0.0.xsd" ?>
<project xmlns="http://maven.apache.org/POM/4.0.0" >
	
</project>
```

Open the completion and you should see pom completion:

image

and hover:

image

You can add several xml-model and benefit with completion for each grammar (according to the possible elements for xsd) but I don't understand how we could use it with dtd and xsd both.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java [hunk 1]
```diff
@@ -38,12 +40,26 @@ public String onTag(IHoverRequest hoverRequest) throws Exception {
...
+					if (content != null) {
+						return content.getValue();
+					}
+					return null;
+				}
+				// -> Multiple element declaration
+				StringBuilder result = new StringBuilder();
+				for (CMElementDeclaration cmElement : cmElements) {
+					MarkupContent content = XMLGenerator.createMarkupContent(cmElement, hoverRequest);
+					if (content != null) {
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java:58]: shouldn't you add a separator?

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java [hunk 2]
```diff
@@ -56,20 +72,40 @@ public String onAttributeName(IHoverRequest hoverRequest) throws Exception {
...
+			// There are some element declarations
+			// -> One element declaration
+			if (cmElements.size() == 1) {
+				return getAttributeNameHover(cmElements.get(0), attributeName, hoverRequest);
+			}
+			// -> Multiple element declaration
+			StringBuilder result = new StringBuilder();
+			for (CMElementDeclaration cmElement : cmElements) {
+				String content = getAttributeNameHover(cmElement, attributeName, hoverRequest);
+				if (content != null) {
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java:90]: shouldn't you add a separator?

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java [hunk 3]
```diff
@@ -87,23 +123,42 @@ public String onAttributeValue(IHoverRequest hoverRequest) throws Exception {
...
+			// There are some element declarations
+			// -> One element declaration
+			if (cmElements.size() == 1) {
+				return getAttributeValueHover(cmElements.get(0), attributeName, attributeValue, hoverRequest);
+			}
+			// -> Multiple element declaration
+			StringBuilder result = new StringBuilder();
+			for (CMElementDeclaration cmElement : cmElements) {
+				String content = getAttributeValueHover(cmElement, attributeName, attributeValue, hoverRequest);
+				if (content != null) {
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java:142]: shouldn't you add a separator?

**Comment by angelozerr** [comment]
@fbricon I managed now aggregation with proper mean but I need to write test now.

To try it you can open the grammar.xml (which defines 2 xml-model with a xsd and a dtd) that I have included in this PR and you hover the `from-xsd` element which is declared in the dtd and xsd both.

You should see:

image

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java (hunk 3 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java]: Indeed, I fixed it.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java (hunk 2 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java]: Indeed, I fixed it.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelHoverParticipant.java]: Indeed, I fixed it.

**Comment by angelozerr** [comment]
@fbricon please note you have in this case duplicate of completion item:

image

I keep like this, because the apply completion could be different from the xsd or dtd (ex: xsd defines required attributes, the apply will insert from-xsd element with attributes. The from-xsd declared in the dtd doesn't declare attributes, it generate xsd-from without attributes).

**Comment by angelozerr** [comment]
Test are written for hover aggregation.

**Comment by xorye** [comment]
@angelozerr I have an NPE when I have this xml:
```
<?xml-model href="grammar.xsd"?>
<project>
</project>
```

with this xsd:
```
<element name='project' xmlns='http://www.w3.org/2001/XMLSchema'>
  <element name='element1'>
    <complexType>
      <sequence>
        <element name='element1.1' type='string' />
        <element name='element1.2' type='string' />
      </sequence>
    </complexType>
  </element>
</element>
```

is my xsd file missing something?

This is the error stack trace:
```
[Error - 2:08:58 p.m.] May 20, 2020 02:08:58 org.eclipse.lemminx.extensions.contentmodel.participants.diagnostics.XMLValidator doDiagnostics()
Message: Unexpected XMLValidator error
java.lang.NullPointerException
    at org.apache.xerces.impl.xs.traversers.XSDocumentInfo.<init>(Unknown Source)
    at org.apache.xerces.impl.xs.traversers.XSDHandler.constructTrees(Unknown Source)
    at org.apache.xerces.impl.xs.traversers.XSDHandler.parseSchema(Unknown Source)
    at org.apache.xerces.impl.xs.XMLSchemaLoader.loadSchema(Unknown Source)
    at org.apache.xerces.impl.xs.XMLSchemaValidator.findSchemaGrammar(Unknown Source)
    at org.apache.xerces.impl.xs.XMLSchemaValidator.handleStartElement(Unknown Source)
    at org.apache.xerces.impl.xs.XMLSchemaValidator.startElement(Unknown Source)
    at org.eclipse.lemminx.extensions.xerces.xmlmodel.XMLModelSchemaValidator.startElement(XMLModelSchemaValidator.java:97)
    at org.eclipse.lemminx.extensions.xerces.xmlmodel.XMLModelHandler.startElement(XMLModelHandler.java:191)
    at org.apache.xerces.impl.XMLNSDocumentScannerImpl.scanStartElement(Unknown Source)
    at org.apache.xerces.impl.XMLNSDocumentScannerImpl$NSContentDispatcher.scanRootElementHook(Unknown Source)
    at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source)
    at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)
    at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
    at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
    at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
    at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
    at org.eclipse.lemminx.extensions.contentmodel.participants.diagnostics.XMLValidator.parseXML(XMLValidator.java:107)
    at org.eclipse.lemminx.extensions.contentmodel.participants.diagnostics.XMLValidator.doDiagnostics(XMLValidator.java:93)
    at org.eclipse.lemminx.extensions.contentmodel.participants.diagnostics.ContentModelDiagnosticsParticipant.doDiagnostics(ContentModelDiagnosticsParticipant.java:48)
    at org.eclipse.lemminx.services.XMLDiagnostics.doExtensionsDiagnostics(XMLDiagnostics.java:85)
    at org.eclipse.lemminx.services.XMLDiagnostics.doDiagnostics(XMLDiagnostics.java:53)
    at org.eclipse.lemminx.services.XMLLanguageService.doDiagnostics(XMLLanguageService.java:153)
    at org.eclipse.lemminx.services.XMLLanguageService.publishDiagnostics(XMLLanguageService.java:162)
    at org.eclipse.lemminx.XMLTextDocumentService.validate(XMLTextDocumentService.java:464)
    at org.eclipse.lemminx.XMLTextDocumentService.lambda$triggerValidationFor$23(XMLTextDocumentService.java:457)
    at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670)
    at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646)
    at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:457)
    at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
    at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
    at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
    at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by xorye** [comment]
Also for reference, could you provide `grammar.xsd` and `grammar.dtd` from this screenshot, please?
image

**Comment by angelozerr** [comment]
> Also for reference, could you provide grammar.xsd and grammar.dtd from this screenshot, please?

Those files belong to this PR (in the test xml-model folder).

But here their content:

 * `grammar.dtd`:

```xml
<!ELEMENT grammar (#PCDATA | from-dtd | from-xsd)* >
<!ELEMENT from-dtd (#PCDATA)>
<!-- DTD documentation for from-xsd -->
<!ELEMENT from-xsd (#PCDATA)>
```

 * `grammar.xsd`:

```xml
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
	<xs:element name="grammar">
		<xs:complexType>
			<xs:sequence>
				<xs:element name="from-xsd" >
				<xs:annotation>
					<xs:documentation>XSD documentation for from-xsd</xs:documentation>
				</xs:annotation>
				</xs:element>
			</xs:sequence>
		</xs:complexType>
	</xs:element>
</xs:schema>
```

 * `grammar.xml`:

```xml
<?xml-model href="grammar.dtd" ?>
<?xml-model href="grammar.xsd" ?>
<grammar>
	<from-xsd></from-xsd>	
</grammar>
```

**Comment by angelozerr** [comment]
> @angelozerr I have an NPE

The problem is that XSD must start with schema and not element.

```xml
<schema  xmlns='http://www.w3.org/2001/XMLSchema'>
  <element name='element1'>
    <complexType>
      <sequence>
        <element name='element1.1' type='string' />
        <element name='element1.2' type='string' />
      </sequence>
    </complexType>
  </element>
</schema>
```

This NPE is not linkedto this issue but it's a bug of Xerces. If you open the XSD file, you will see the same error. Please create an issue in LemMinx with this XSD which doesn't start with schema. I'm afraid that we cannot do something since it's a bug of Xerces.

**Comment by xorye** [comment]
> Those files belong to this PR (in the test xml-model folder).

Oops sorry, I have missed that the first time around.

The PR is working great on my end, I have also tested it with attribute tag/value completion from an external XSD and DTD

**Comment by angelozerr** [comment]
> The PR is working great on my end, I have also tested it with attribute tag/value completion from an external XSD and DTD

Thanks for the review!

**Pull request merged**